docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,176 @@
+
+- Non-versioned network APIs: Internal network APIs, such as the internal API
+  used to drive the Flow web UI, are not subject to backwards compatibility
+  guarantees.
+
+- Undocumented behavior: relying on undocumented behavior may break between
+  minor releases. 
+
+- Upstream dependencies: part of the public API of Grafana Agent may directly
+  expose the public API of an upstream dependency. In these cases, if an
+  upstream dependency introduces a breaking change, we may be required to make
+  a breaking change to our public API as well.   
+
+- Other telemetry data: Metrics, logs, and traces may change between releases.
+  Only telemetry data which is used in official dashboards is protected under
+  backwards compatibility.
+
+### Avoiding breaking changes 
+
+Backwards compatibility guarantees introduce new challenges while writing code,
+as it means finding creative ways to iterate without making breaking changes
+immediately. Here are some example strategies one could use to avoid making a
+breaking change:
+
+- Keep things internal: The easiest route is to just not expose things
+  publicly. For example, put Go code in `internal/` packages, or use unexported
+  methods or types.
+
+- Rely on deprecations: If a publicly exposed thing must be changed, consider
+  deprecating it.
+
+  - Instead of removing something, consider making it a no-operation and adding
+    a deprecation warning. 
+
+  - Instead of replacing something (Go methods, Flow components, Flow attribute
+    names), consider introducing the new approach and deprecating the old
+    approach. 
+
+A list of deprecations will accumulate over the course of a major version's
+lifecycle. This list must be tracked so that deprecated functionality can be
+removed with the next major version. 
+
+### Avoiding major release burnout 
+
+As a new major release implies a user must put extra effort into upgrading, it
+is possible to burn out users by releasing breaking changes too frequently. 
+
+We will attempt to limit new major versions no more than once every 12 calendar
+months. This means that if Grafana Agent 1.0 was hypothetically released on
+August 4th, Grafana Agent 2.0 should not be released until at least August 4th
+of the following year. This is best-effort; if a new major release is required
+earlier, then one should be published. 
+
+Major releases should be aligned with breaking changes to the public API and
+not used as a way to hype a release. If hyping releases is required, there
+should be a version split between the API version of Grafana Agent and a
+project version (such as API v1.5, project version 2023.0).   
+
+### Supporting previous major releases
+
+When a new major release is published, the previous major release should
+continue to receive security and bug fixes for a set amount of time.
+Announcement of a new major release should be coupled with a minimum Long-Term
+Support (LTS) period for the previous major release. For example, we may choose
```

Comment:
Do we want to expand the definition of LTS in terms on number of months or this is out of the scope of this document?

Reply:
I try to keep it out of scope in the last paragraph:

> The support timeframe for an LTS version is not fixed, and may change between
major releases. For example, version 0.X may receive at least 12 months of LTS,
while version 1.X may receive at least 4 months of LTS. Project maintainers
will need to decide how long to support previous major versions based on the
difficulty for upgrading to the latest major version.